feat: Implement is_valid algorithm for polyhedral surfaces

vissarion opened the pull request:
This PR introduces support for the `is_valid` algorithm for polyhedral surface geometries. A polyhedral_surface is a 3D composite geometry formed by a collection of polygonal faces that meet at shared edges. The implementation follows the OGC Simple Features Specification and introduces structural and topological checks for polyhedral surfaces.

The following properties should hold (following OGC standard):

- Contiguity: All polygon patches must form a connected surface via shared boundary segments.
- Shared Edges: Shared boundaries must be finite LineStrings used in at most two polygons.
- Orientation: Shared boundaries must be traversed in opposite directions by adjacent polygons; global orientation must be consistent.

Six new validity failure types has been introduced that reflect the description of the standard. 

Below there is a matrix with valid and invalid cases with visualizations. Those cases (and a few more) are included in the unit tests of this PR. 

| Name                                               | Geometry                                                                                                                                                                                                                                     | Figure |
|----------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------|
| valid_surface                                      | POLYHEDRALSURFACE(((0 0 0,0 1 0,1 1 0,1 0 0,0 0 0)),((0 0 0,0 0 1,0 1 1,0 1 0,0 0 0)),((0 0 0,1 0 0,1 0 1,0 0 1,0 0 0)),((1 1 1,0 1 1,0 0 1,1 0 1,1 1 1)),((1 1 1,1 0 1,1 0 0,1 1 0,1 1 1)),((1 1 1,1 1 0,0 1 0,0 1 1,1 1 1))) | ![Screenshot from 2025-05-21 16-52-54](https://github.com/user-attachments/assets/0ddb2629-bf4b-49ab-9992-ddfba7b1ee18) |
| valid_surface_non_convex                          | POLYHEDRALSURFACE(((0 0 0,0 1 0,1 1 0,1 0 0,0 0 0)),((0 0 0,0 0 1,0 1 1,0 1 0,0 0 0)),((0 0 0,1 0 0,1 0 1,0 0 1,0 0 0)),((1 1 1,0 1 1,0 0 1,1 0 1,1 1 1)),((1 1 1,1 0 1,1 0 0,1 1 0,1 .5 .5,1 1 1)))               | ![Screenshot from 2025-05-21 16-53-30](https://github.com/user-attachments/assets/954adfe2-efb4-413e-9758-fd55c65c3d9e) |
| valid_surface_holes                                | POLYHEDRALSURFACE(((0 0 0,0 1 0,1 1 0,1 0 0,0 0 0),(0.2 0.2 0,0.2 0.5 0,0.5 0.5 0,0.5 0.2 0,0.2 0.2 0)),((0 0 0,0 0 1,0 1 1,0 1 0,0 0 0)),((0 0 0,1 0 0,1 0 1,0 0 1,0 0 0)),((1 1 1,0 1 1,0 0 1,1 0 1,1 1 1)),((1 1 1,1 0 1,1 0 0,1 1 0,1 1 1)),((1 1 1,1 1 0,0 1 0,0 1 1,1 1 1))) | ![Screenshot from 2025-05-21 16-55-21](https://github.com/user-attachments/assets/72a00aa7-cf2f-4249-9805-21cbb63c0f47)  |
| invalid_surface_non_planar_face                    | POLYHEDRALSURFACE(((0 0 0,1 0 0,0 1 0,0 0 1,0 0 0)),((1 0 0,1 1 0,0 1 0,1 0 0)))                                                                                                                        | ![Screenshot from 2025-05-21 17-20-01](https://github.com/user-attachments/assets/f0ac921c-6bbb-4d00-b3fd-36267e1d702d)   |
| invalid_surface_collinear_points                   | POLYHEDRALSURFACE(((1 0 0,1 1 0,0 1 0,1 0 0)),((0 0 0,1 0 0,2 0 0,0 0 0)))                                                                                                                              |  ![Screenshot from 2025-05-21 17-10-21](https://github.com/user-attachments/assets/0cbc3240-3088-44ba-b90d-bdfd87ce296c)     |
| invalid_surface_few_points                         | POLYHEDRALSURFACE(((1 0 0,1 1 0,0 1 0,1 0 0)),((0 0 0,1 0 0,0 0 0)))                                                                                                                                    |  ![Screenshot from 2025-05-21 17-11-01](https://github.com/user-attachments/assets/160455b0-b0a6-4393-9ff8-1e9caf368e29)     |
| invalid_surface_invalid_intersection_boundary_interior | POLYHEDRALSURFACE(((0 0 0,1 0 0,0 0 1,0 0 0)),((0 1 0,0.5 -0.5 0.5,1 1 0,0 1 0)))                                                                                                                   | ![Screenshot from 2025-05-21 17-11-20](https://github.com/user-attachments/assets/7805e68c-d576-414d-90d7-999d1ff319f1)      |
| invalid_surface_invalid_intersection_boundary_boundary | POLYHEDRALSURFACE(((0 0 0,1 0 0,1 0 1,0 0 1,0 0 0)),((0 1 0,0 -1 1,1 -1 1,1 1 0,0 1 0)))                                                                                                               | ![Screenshot from 2025-05-21 17-11-36](https://github.com/user-attachments/assets/abe12aea-b14c-480d-8a46-86f47abc980a)      |
| invalid_surface_invalid_intersection_in_common_vertex | POLYHEDRALSURFACE(((0 0 0,1 0 0,1 0 1,0 0 1,0 0 0)),((0 1 0,1 0 1,1 1 0,0 1 0)))                                                                                                                      |  ![Screenshot from 2025-05-21 17-11-57](https://github.com/user-attachments/assets/d702ef12-04b6-4d83-8528-172449472010)     |
| invalid_surface_invalid_intersection_partial_edge   | POLYHEDRALSURFACE(((0 0 0,1 0 0,1 0 1,0 0 1,0 0 0)),((0 1 0,0.5 0 1,1 0 1,1 1 0,0 1 0)))                                                                                                                | ![Screenshot from 2025-05-21 17-12-11](https://github.com/user-attachments/assets/a962160f-af17-40b1-85a2-357d514a97b1)      |
| invalid_surface_disconnected_polygons              | POLYHEDRALSURFACE(((0 0 0,1 0 0,0 1 0,0 0 0)),((1 1 0,2 1 0,2 2 0,1 2 0,1 1 0)))                                                                                                                      |  ![Screenshot from 2025-05-21 17-19-31](https://github.com/user-attachments/assets/909293eb-e24a-4593-8998-8eb6cbf34365)     |
| invalid_surface_invalid_intersection_vertex_edge    | POLYHEDRALSURFACE(((0 0 0,1 0 0,0 1 0,0 0 0)),((0.5 0.5 0,2 1 0,2 2 0,1 2 0,0.5 0.5 0)))                                                                                                                | ![Screenshot from 2025-05-21 17-12-52](https://github.com/user-attachments/assets/67035568-5a18-433d-b450-6906c43465a6)      |
| invalid_surface_invalid_intersection_vertex_vertex  | POLYHEDRALSURFACE(((0 0 0,1 0 0,0 1 0,0 0 0)),((1 0 0,2 1 0,2 2 0,1 2 0,1 0 0)))                                                                                                                        | ![Screenshot from 2025-05-21 17-13-07](https://github.com/user-attachments/assets/ea1c1288-3239-429b-acaa-744377128b16)       |
| invalid_surface_invalid_intersection_parallel_edges | POLYHEDRALSURFACE(((0 0 0,1 0 0,0 1 0,0 0 0)),((0 1 0,0.5 0.5 0,1 0 0,2 1 0,2 2 0,1 2 0,0 1 0)))                                                                                                        | ![Screenshot from 2025-05-21 17-13-24](https://github.com/user-attachments/assets/c8d158f6-ff5f-4ee9-920c-b8055502eeea)       |
| invalid_surface_invalid_intersection_operlapping_faces | POLYHEDRALSURFACE(((0 0 0,1 0.5 0,0.5 1 0,0 0 0)),((0 1 0,1 0 0,2 1 0,2 2 0,1 2 0,0 1 0)))                                                                                                           | ![Screenshot from 2025-05-21 17-13-45](https://github.com/user-attachments/assets/5f488f7a-464e-4782-8767-4ac2718782bd)      |
| invalid_surface_invalid_intersection_operlapping_faces2 | POLYHEDRALSURFACE(((0 0 0,1 0 0,0.5 1 0,0 0 0)),((0 1 0,1 0 0,2 1 0,2 2 0,1 2 0,0 1 0)))                                                                                                             | ![Screenshot from 2025-05-21 17-14-04](https://github.com/user-attachments/assets/ba5d7bad-04a1-4fe7-b1b5-b0a819b09b08)      |
| invalid_surface_invalid_face                        | POLYHEDRALSURFACE(((0 0 0,1 0 0,0 1 0,0 0 0)),((0 1 0,1 0 0,2 2 0,2 1 0,1 2 0,0 1 0)))                                                                                                                  | ![Screenshot from 2025-05-21 17-14-25](https://github.com/user-attachments/assets/61c2c2e4-0960-4fde-9de5-e05b467a66ab)      |
| invalid_surface_non_manifold_edge                  | POLYHEDRALSURFACE(((0 0 0,1 0 0,0 1 0,0 0 0)),((0 1 0,1 0 0,2 1 0,2 2 0,1 2 0,0 1 0)),((0 1 0,1 0 0,1 1 1,0 1 0)))                                                                                       | ![Screenshot from 2025-05-21 17-14-43](https://github.com/user-attachments/assets/47c1fbeb-f99c-4a4e-a099-14d07ca69af3)      |

Two unit tests are failing and are commented out. Those are dependent on https://github.com/boostorg/geometry/issues/1406

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+                auto prev = (it == boost::begin(ring)) ? boost::end(ring) - 1 : boost::prior(it);
+                if (boost::geometry::equals(point2, *prev))
+                {
+                    // valid intersection and inconsistent orientation
```
Should this be ~~in~~consistent?

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+    };
+
+    // Check if two vectors are linearly dependent (collinear)
+    static bool are_linearly_dependent(vec3 const& a, vec3 const& b)
+    {
+        return qvm::mag_sqr(qvm::cross(a, b)) < tolerance * tolerance;
+    }
+
+    // Try to find 3 non-collinear points
+    static bool find_non_collinear_points(polygon_3d const& polygon, vec3& base, vec3& v1, vec3& v2)
+    {
+        for (auto pit1 = points_begin(polygon); pit1 != points_end(polygon); ++pit1)
+        {
+            for (auto pit2 = pit1; pit2 != points_end(polygon); ++pit2)
+            {
+                for (auto pit3 = pit2; pit3 != points_end(polygon); ++pit3)
+                {
+                    set_vec_from_point::apply(*pit1, base);
+                    set_vec_from_point::apply(*pit2, v1);
+                    set_vec_from_point::apply(*pit3, v2);
+                    if (!are_linearly_dependent(v1 - base, v2 - base))
+                    {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false; // All points are collinear
+    }
```
I think by the library design philosophy and also practically, if one wants to be able to plug in a robust computation later, this is a cartesian strategy.

vissarion replied:
Indeed, a lot of computations here are coordinate system spesific and should be moved in strategies. On one hand, I think that polyhedral surface can only be cartesian. But on the other hand, predicates should be moved to stategies anyways for robustness reasons (as you said).

vissarion replied:
That reduces the computation to orientation, great. I was thinking using the `project_point_to_plane_2d` to only do one orientation test but I think projecting to coordinate planes as you suggested is better since it does not involve any computations (substitutions, dot products etc). 
I change this part to use the side strategy provided by the umbrella strategy used in is_valid. This uses the side_by_triangle strategy as the default but the user can change it (as in many other cases in the library). At this point I think this is better then using a creating a new umbrella strategy that has side_rounded_input as a side strategy. We can think of substituting the side_by_triangle by side_rounded_input in general or in as many algorithms as possible.

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+    // Check if two vectors are linearly dependent (collinear)
+    static bool are_linearly_dependent(vec3 const& a, vec3 const& b)
+    {
+        return qvm::mag_sqr(qvm::cross(a, b)) < tolerance * tolerance;
```
We may want some factor on this bound, I will recheck.

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+        // Check if the polygon has been projected before
+        std::ostringstream oss;
+        oss << boost::geometry::wkt(polygon);
+        if (m_polygon_projections.find(oss.str()) != m_polygon_projections.end())
```
This seems rather expensive aside from pulling the wkt dependency into this. Can we use the indices of the polygons in their original datastracture as integer keys instead? (I understand that all boost geometry concepts require random access ranges).

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+        {
+            for (auto it2 = boost::begin(surface); it2 != boost::end(surface); ++it2)
+            {
+                if (std::distance(boost::begin(surface), it1) !=
+                    std::distance(boost::begin(surface), it2))
```
Is this different from it1 != it2?

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+    {
+        boost::ignore_unused(strategy);
+        bool disconnected = true;
+        // TODO: improve perofmance by using a spatial index
```
Typo: performance

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+        set_vec_from_point<point_3d, 0, 3>;
+
+    using point_2d = boost::geometry::model::point
+        <CT, 2, boost::geometry::cs::cartesian>;
```
Are polyhedral surfaces always cartesian (I didn't see a corresponding constraint in the concept check https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/geometries/concepts/polyhedral_surface_concept.hpp but I may have missed it) and if not should this be generic or is_valid constrained for now?

barendgehrels replied:
I think it's a valid constraint at least for now

vissarion replied:
I added that as a concept constraint together with the constraint on dimension (==3).

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+        oss << boost::geometry::wkt(polygon);
+        if (m_polygon_projections.find(oss.str()) != m_polygon_projections.end())
+        {
+            auto& projection = m_polygon_projections[oss.str()];
```
Minor: This seems to be only copied from, so maybe it can be declared const.

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+                }
+            }
+            // Project the polygon onto the 2D plane defined by v1 and v2
+            for (auto& point : ring)
```
This point seems to only be read from.

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+        // Compute the intersection point
+        vec3 seg_dir = p2 - p1;
+        CT denom = dot(normal, seg_dir);
+        assert(denom != 0); // Ensure the segment is not parallel to the plane
```
Is this necessarily a bug where the program should be halted in a build with asserts? Seems like something that could also happen with e.g. invalid coordinate inputs (e.g. denormal small coords so that product is zero, since normal is not normalised afaics) that should not be a halt but the check returning an is_valid failure.

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+                vec3 vi;
+                set_vec_from_point::apply(*pit, vi);
+                auto vol = qvm::dot(normal, vi)- dot(normal, base);
+                if (std::abs(vol) > tolerance)
```
This may be to narrow for e.g. float coordinates in the range of thousand, I think, the bound should probably be scaled to the input magnitude, I will revisit this later when I have more time.

tinko92 replied:
This is orient3d (see e.g. https://www.[cs.cmu.edu/afs/cs/project/quake/public/code/predicates.c](https://www.cs.cmu.edu/afs/cs/project/quake/public/code/predicates.c)), i.e. the 3D generalisation of side, a determinant involving `*pit`, `base`, `v1`, `v2`. Bound for a tolerant version can be derived the same way as for side_rounded_input, I'll try to do it by the end of the week. Calling orient3d for every point may look more expensive, but I think the compiler will know to pull most unchanging subexpressions out of the loop (same as dot(normal, base) could be pulled out of the loop).

tinko92 replied:
I thought about it. I think, we want

```math
\begin{vmatrix}
    a_x-d_x & a_y -d_y & a_z - d_z \\
    b_x-d_x & b_y-d_y & b_z-d_z \\
    c_x-d_x & c_y-d_y & c_z-d_z
\end{vmatrix}
```

Since only one point changes (let's say $a=\text{vi}$), it makes sense to take the first row as cofactors and the minors will be the elements of $n$ (if all points, including $a$ were translated by base before). Then the determinant is 

$(a_x-d_x) \times ((b_y-d_y) \times (c_z - d_z) - (b_z - d_z) \times (c_y - d_y)) - (a_y - d_y) \times \ldots$

Then, I think, the tolerance (based on https://zenodo.org/records/7539355 with rounded input rules, rounding the second order part up because it doesn't matter for this use case):

```math
12\epsilon \times \left(\left(\left\lvert a_x\right\rvert  + \left\lvert d_x\right\rvert \right) \times \left(\left(\left\lvert b_y\right\rvert  + \left\lvert d_y\right\rvert \right) \times \left(\left\lvert c_z\right\rvert  + \left\lvert d_z\right\rvert \right) + \left(\left\lvert b_z\right\rvert  + \left\lvert d_z\right\rvert \right) \times \left(\left\lvert c_y\right\rvert  + \left\lvert d_y\right\rvert \right)\right) + \left(\left\lvert a_y\right\rvert  + \left\lvert d_y\right\rvert \right) \times \left(\left(\left\lvert b_x\right\rvert  + \left\lvert d_x\right\rvert \right) \times \left(\left\lvert c_z\right\rvert  + \left\lvert d_z\right\rvert \right) + \left(\left\lvert b_z\right\rvert  + \left\lvert d_z\right\rvert \right) \times \left(\left\lvert c_x\right\rvert  + \left\lvert d_x\right\rvert \right)\right) + \left(\left\lvert a_z\right\rvert  + \left\lvert d_z\right\rvert \right) \times \left(\left(\left\lvert b_x\right\rvert  + \left\lvert d_x\right\rvert \right) \times \left(\left\lvert c_y\right\rvert  + \left\lvert d_y\right\rvert \right) + \left(\left\lvert b_y\right\rvert  + \left\lvert d_y\right\rvert \right) \times \left(\left\lvert c_x\right\rvert  + \left\lvert d_x\right\rvert \right)\right)\right)
```

where $\epsilon$ is **half** of numerical_limits<CT>::epsilon(). This is the same structure as the expanded determinant. If I made no mistake, this will guarantee that the tolerance is sufficiently large such that for all input points that are correctly rounded approximations of actually collinear points, it will find them coplanar.

And this would be the equivalent of side_rounded for planes.

tinko92 reviewed (commented):
I noted some points inline, mostly minor but importantly (imho) the question of separating some computations into strategy.

A minor point, this PR should probably include the removal of  https://github.com/boostorg/geometry/blob/d2e5137f4070d8bbcd8ab150ad4fe42713c8ceb3/include/boost/geometry/geometries/polyhedral_surface.hpp#L47

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,462 @@
+                projected_polygon.outer().push_back(projected_p);
+            }
+            // Compute the orientation of the polygon and reverse it if necessary
+            if (boost::size(ring) >= 3)
+            {
+                point_2d p0 = projected_polygon.outer()[0];
+                point_2d p1 = projected_polygon.outer()[1];
+                point_2d p2 = projected_polygon.outer()[2];
+
+                auto side = strategy.side().apply(p0, p1, p2);
+
+                auto const& orientation = boost::geometry::point_order<polygon_2d>::value;
+                if ((side > 0 && orientation == boost::geometry::clockwise) ||
+                    (side < 0 && orientation == boost::geometry::counterclockwise))
+                {
+                    boost::geometry::reverse(projected_polygon);
+                }
```
Is this implictly assuming that 
1) The polygon is convex? If not, three points may be insufficient to determine its orientation. It is not obvious to me from what I can find on the simple feature spec that it has to be.
2) Its first three points are not collinear? They may be collinear, even for a valid, convex polygon.

vissarion replied:
Note that the correct orientation is tested by is_valid call to projected polygon below.

barendgehrels commented:
Looks great! I will review it next Wednesday or Thursday.

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+        : m_polygon_projections(num_of_polygons)
+    {}
+
+    using CT = typename coordinate_type<PolyhedralSurface>::type;
```
Suggestion, `coor_t`? `CT` looks like a template parameter

vissarion replied:
I agree, I changed it to `coordinate_type` which seems more common as a use in the library.

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+    using vec3 = boost::qvm::vec<CT, 3>;
+    using polygon_3d = typename boost::range_value<PolyhedralSurface>::type;
+    using ring_3d = typename boost::geometry::ring_type<polygon_3d>::type;
+    using point_3d = typename boost::range_value<ring_3d>::type;
```
Why not `point_type` ? It delivers the same but we usually use that. You don't need the ring then (but maybe it's needed anyway)

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+        {
+            for (auto it2 = boost::begin(surface); it2 != boost::end(surface); ++it2)
+            {
+                if (it1 != it2)
```
Suggestion, `if (it1 == it2) continue`

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+                             Strategy const& strategy)
+    {
+        boost::ignore_unused(strategy);
+        bool disconnected = true;
```
Suggestion, call it `connected = false` and set it later to `true`. For me that would read a bit better

vissarion replied:
This is also removed. There is a similar structure/logic in `check_polygon_polygon_intersection` where I used your naming approach.

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+        return intersection_type::none;
+    }
+
+    // Helper function to check if any segment of a ring intersects a polygon
```
Do we really support polygons? Or rings? It's not completely clear to me.
This is checking the outer ring only?

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+                                            Strategy const& strategy)
+    {
+        auto const& ring1 = boost::geometry::exterior_ring(polygon1);
+        for (auto it = boost::begin(ring1); it != boost::end(ring1) - 1; ++it)
```
Suggestion: `it + 1 != boost::end(ring1)` to avoid empty ring problems

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+        }
+        else
+        {
+            auto const& ring = boost::geometry::exterior_ring(polygon);
```
So this block does really the projection. Nice.
I think it's good to extract it from this class - and even to another source.
Such that it is useful on its own, and can be unit tested on its own.

barendgehrels replied:
But exactly that (shared variables) will makes it harder to extract it later as well...

A generic polygon projection from 3d to 2d plane would be great. We could then immediately implement the surface-area algorithm on a 3d surface... 

It would be a generic algorithm, every patch (surface) would be a separate 2D polygon. Or is it different than what you do here (I couldn't look into all the details yet).

Anyway - I'm fine having it as a next PR - I'm only afraid that it would not really safe time.

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+                    return intersection_type::invalid;
+                }
+            }
+            // Project the polygon onto the 2D plane defined by v1 and v2
```
Looks interesting. I'm not experienced in this so cannot really comment.

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+        // Check if the segment intersects the plane of the polygon
+        CT dot_product = dot(normal, base);
+        CT sign1 = dot(normal, p1) - dot_product;
+        CT sign2 = dot(normal, p2) - dot_product;
```
This can all be const

tinko92 commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+        {
+            auto const& ring = boost::geometry::exterior_ring(polygon);
+            constexpr bool is_ring_open = boost::geometry::closure<ring_3d>::value
+                == boost::geometry::open;
```
Since this is constexpr, this is only done once at compile time, there may be merit to scoping it for separation of logic,.

barendgehrels replied:
Agreed - but the whole list of types is also at the top (not scoped). This one fits there.

vissarion commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+                    }
+                    else if (intersection == intersection_type::valid)
+                    {
+                        disconnected = false;
```
You are right. Currently, the check only checks that there is no disconnected polygon. More work is needed to check for connectedness of the surface.

vissarion commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+            typename boost::geometry::model::polygon<point_2d, false>
+        >::type;
+    using ring_2d = typename boost::geometry::ring_type<polygon_2d>::type;
+    using segment_2d = typename boost::geometry::model::segment<point_2d>;
```
segment_2d is not used any more so I removed. Ring types we can get them from `std::dectype` if you prefer it, I am ok with those as is.

barendgehrels replied:
It's fine. But did you push already? I don't see the updates...

vissarion commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+                disconnected = false; // Found valid intersection
+            }
+        }
+        return true; // Either no intersection or valid intersection
```
Now all functions in `valid_intersection` return `intersection_type`, I think it is easier to follow the code.

vissarion commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+        polygon_2d projected_polygon;
+
+        // Check if the polygon has been projected before
+        if (m_polygon_projections[polygon_index].projected_polygon.outer().size() > 0)
```
I used the `boost::geometry::is_empty` function.

vissarion commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,453 @@
+            {
+                vec3 p;
+                set_vec_from_point::apply(point, p);
+                auto projected_p = project_point_to_plane_2d(v1, v2, p);
```
I changes to `projected_polygon.outer().emplace_back(project_point_to_plane_2d(v1, v2, p));`
that should avoid copies

vissarion commented:
I refactor the code addressing almost all of the comments. Some things are left as todo for next PR(s). What is missing from this PR is the predicates (places in the code that use tolerance) that should be written as orientation checks implemented in strategies. I will work on it in the next days.

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+    };
+
+    template <int I, int J, typename Strategy>
+    static bool affinely_independent(point_3d const& p1, point_3d const& p2, point_3d const& p3,
```
Suggestion for name: `is_affinely_independent`

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+        return strategy.side().apply(p1_2, p2_2, p3_2) != 0;
+    }
+
+    // Try to find 3 non-collinear points
```
in any plane?

vissarion replied:
3 non-colinear points will define a plane. Later in the code we check if the rest of the points of the polygon lay on this plane.

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+                {
+                    if (affinely_independent<0, 1>(*pit1, *pit2, *pit3, strategy)
+                        || affinely_independent<0, 2>(*pit1, *pit2, *pit3, strategy)
+                        || affinely_independent<1, 2>(*pit1, *pit2, *pit3, strategy))
```
Suggestion, add comment `One three-point-set is already sufficient, with such three independent points a plane can be constructed` (or similar).

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+        return false; // All points are collinear
+    }
+
+    static point_2d project_point_to_plane_2d(vec3 const& v1, vec3 const& v2, vec3 const& p)
```
Without comments, and drawings, and much experience, it's hard to follow this...

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+        auto ring = boost::geometry::exterior_ring(polygon);
+        if (boost::geometry::closure<ring_2d>::value == boost::geometry::closed)
+        {
+            ring.pop_back(); // Remove the closing point to make it open
```
We have also an iterator doing this automatically. It closes it automatically (which is similar - but you can then always skip the last point).
That avoids the need for a copy of the ring.

vissarion replied:
Thanks, I will use `closing_iterator`.

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+            {
+                if (boost::geometry::equals(point1, point2))
+                {
+                    visitor.template apply<failure_duplicate_points>();
```
Can't this check be made right at the beginning, before even iterating?
What are point1 and point2? Is it a segment?
Can we then rename them to `segment_point1` or just `s1` with a comment?

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+        return point_2d{qvm::dot(v1, p), qvm::dot(v2, p)};
+    }
+
+    template <typename VisitPolicy>
```
So what does it do exactly? It finds if the segment point1-point2 overlaps with any segment of the outer ring of the polygon?

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+        else
+        {
+            // TODO: Support interior rings
+            ring_3d const ring = boost::geometry::exterior_ring(polygon);
```
It's a copy...

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+    using ring_3d = typename boost::geometry::ring_type<polygon_3d>::type;
+    using point_3d = typename boost::geometry::point_type<ring_3d>::type;
+    using segment_3d = typename boost::geometry::model::segment<point_3d>;
+    using set_vec_from_point = typename strategy::transform::detail::matrix_transformer::
```
The name is not 100% clear to me.

But (I see now) we have that already for a long time.
It sets a vector from a point?
Anyway, the `using` clause is fine then...

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+        return false;
+    }
+
+    // Check if a segment intersects a polygon in 3D space
```
So also here `point1` and `point2` form a segment, it would be good to make that clear.

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+        vec3 p1;
+        vec3 p2;
+        set_vec_from_point::apply(point1, p1);
+        set_vec_from_point::apply(point2, p2);
```
Why not `v1` and `v2` here, if these are vectors? 
And then below you have another set of `v1` and `v2`, it's not clear to me...

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,501 @@
+        // Check if the segment intersects the plane of the polygon
+        coordinate_type const dot_product = dot(normal, base);
+        coordinate_type const sign1 = dot(normal, p1) - dot_product;
+        coordinate_type const sign2 = dot(normal, p2) - dot_product;
```
The `p1` and `p2` are calculated long before...
Would it be possible to move that to here? (And rename them to something vector-like)?

vissarion commented:
@barendgehrels thanks for you comments! I addressed most of them. I need to work more on this PR. Apart from the orient3d some other simplifications can be implemented (the unaddressed comments of Barend are related to those simplifications). I will work on it next week.

vissarion commented:
I have updated this PR. Now strategies are used (with 2d and 3d orientation predicates) and there are no constructions on new geometries apart from coordinate projections (that are not that vulnerable to numerical errors). Also a new algorithm for segment-polygon intersection in 3D only using orientation predicates is implemented.

barendgehrels commented on include/boost/geometry/strategies/relate3/cartesian.hpp:
```diff
@@ -0,0 +1,58 @@
+// Contributed and/or modified by Vissarion Fysikopoulos, on behalf of Oracle
+
+// Licensed under the Boost Software License version 1.0.
+// http://www.boost.org/users/license.html
```
AFAIK we don't use this two-line version.
Also, the other file added contains:
```
// Use, modification and distribution is subject to the Boost Software License,
// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)
```

I'm OK with the shorter version if:
* it is used in other libraries too (do you have a reference)
* it is consistent within this PR

and then we can modify it in all our sources (maybe in one go)

vissarion replied:
Thanks for noticing. Actually, there are ~300 files in BG that use the shorter version, while ~1700 that use the longer one. Other boost projects use slightly different versions e.g.
```
Distributed under the Boost Software License, Version 1.0.
(See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
```
used by `hana` and a similar one by `beast`.

So for this PR I will use the longer one and we have to decide how to proceed with all the sources.

barendgehrels replied:
Thanks. And wow... so we've 3 versions indeed, starting with either
```
// Distributed under the Boost Software License, Version 1.0.
// Licensed under the Boost Software License version 1.0.
// Use, modification and distribution is subject to the Boost Software License,
```
and in the whole Boost there are 103 versions (some caused by different way of commenting, but still 50 then)

Actually - why did we start using the version starting with "Licensed"? Because we're the only ones... Was this prescribed by Oracle?

By far the most frequent I see (12411 times) the one you mention with `Distributed under the Boost Software License`

Let's go for that one. But we can have the following lines together (because we use the 100 col. ruler).

I actually also like the ones adding the repository (as done in `url`) and documentation (as done in `unordered`).

So I would propose:
```
// Distributed under the Boost Software License, Version 1.0.
// (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

// Official repository: https://github.com/boostorg/geometry
// Documentation: http://www.boost.org/libs/geometry
```

vissarion replied:
> Actually - why did we start using the version starting with "Licensed"? Because we're the only ones... Was this prescribed by Oracle?

I do not think so, probably it happened accidentally. 

>So I would propose:
```
// Distributed under the Boost Software License, Version 1.0.
// (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

// Official repository: https://github.com/boostorg/geometry
// Documentation: http://www.boost.org/libs/geometry
```
I agree.

barendgehrels commented on include/boost/geometry/strategies/cartesian/side_3d_rounded_input.hpp:
```diff
@@ -0,0 +1,86 @@
+// Boost.Geometry (aka GGL, Generic Geometry Library)
```
we don't need `aka GGL` anymore, in new sources I always omit it.

barendgehrels commented on include/boost/geometry/strategies/cartesian/side_3d_rounded_input.hpp:
```diff
@@ -0,0 +1,86 @@
+// Contributed and/or modified by Vissarion Fisikopoulos, on behalf of Oracle
+
+// This file was modified by Oracle on 2023.
+// Modifications copyright (c) 2023 Oracle and/or its affiliates.
```
Should be `2025`

barendgehrels commented on include/boost/geometry/strategies/cartesian/side_3d_rounded_input.hpp:
```diff
@@ -0,0 +1,86 @@
+        static coor_t const eps = std::numeric_limits<coor_t>::epsilon() / 2;
+        coor_t const det = (p_x - p1_x) * ((p2_y - p1_y) * (p3_z - p1_z) - (p3_y - p1_y) * (p2_z - p1_z))
+                         - (p_y - p1_y) * ((p2_x - p1_x) * (p3_z - p1_z) - (p3_x - p1_x) * (p2_z - p1_z))
+                         + (p_z - p1_z) * ((p2_x - p1_x) * (p3_y - p1_y) - (p3_x - p1_x) * (p2_y - p1_y));
```
Nice... it could profit from some ASCII art

barendgehrels commented:
Cool changes, thanks for the splits - I'll try to finish my review this weekend

tinko92 reviewed (approved):
Thanks for all the work on incorporating the requested changes! This looks good to me. I left some comments on formatting (with diffs, see attached txt files in comments) and compiler warnings, but nothing that would require another review as far as I can see, so I'm marking it as approve.

vissarion commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,648 @@
+namespace detail { namespace is_valid
+{
+
+struct polygon_frontier_type : public std::stack<size_t>
```
It is a leftover, thanks.

vissarion commented:
Thanks for your reviews! I addressed all the comments. 
(However, I think @barendgehrels didn't finish his review yet)

barendgehrels commented on include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp:
```diff
@@ -0,0 +1,650 @@
+// - 2 if the segment intersects the polygon at a vertex
+// - 3 if the segment intersects the polygon at an edge
+// Reference: R.J.Segura, F.R.Feito - An algorithm for determining intersection segment-polygon in
+// 3D, Computers & Graphics 1998
```
👍

barendgehrels commented on include/boost/geometry/algorithms/detail/overlay/graph/assign_counts.hpp:
```diff
@@ -80,22 +80,22 @@ void assign_counts(Turn& turn)
     };
 
-    auto assign_right_incoming_for = [&turn, &assign_for](counts_per_op_t const& op1, counts_per_op_t op2)
+    auto assign_right_incoming_for = [&assign_for](counts_per_op_t const& op1, counts_per_op_t op2)
```
Thanks for the side activity ;-)

barendgehrels commented on include/boost/geometry/strategies/cartesian/side_3d_rounded_input.hpp:
```diff
@@ -0,0 +1,98 @@
+// The following determinant is used to calculate the side:
+//       | (p_x - p1_x)  (p_y - p1_y)  (p_z - p1_z)  |
+// det = | (p2_x - p1_x) (p2_y - p1_y) (p2_z - p1_z) |
+//       | (p3_x - p1_x) (p3_y - p1_y) (p3_z - p1_z) |
```
👍

barendgehrels reviewed (approved):
Big step! Thanks!

tinko92 commented:
Thanks for the quick response. The latest force push broke `b2 test`, looks like a missing `git add test/geometries/polyhedral_surface_fail.cpp`.
